### ISSUE TITLE
[Bug] Cross-cloud calls to eSTS break when region discovery enabled

### ISSUE THREAD

**Issue by agolotin** [AzureAD/microsoft-authentication-library-for-dotnet#2746]
**Logs and network traces**

When issuing a request from Azure Gov cloud to Public identity endpoint we found the following issue:

```
MSAL.Desktop.4.32.1.0.MsalServiceException: 
	ErrorCode: invalid_client
Microsoft.Identity.Client.MsalServiceException: A configuration issue is preventing authentication - check the error message from the server for details. You can modify the configuration in the application registration portal. 
See https://aka.ms/msal-net-invalid-client for details.  
Original exception: AADSTS700023: Client assertion audience claim does not match Realm issuer. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .
Trace ID: fc08600f-2b4b-4f32-aea1-31fcee73fd00
Correlation ID: 6ae91953-06b9-4292-94ab-b5149ff0f5cd
Timestamp: 2021-06-28 16:31:43Z
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.HandleTokenRefreshError(MsalServiceException e, MsalAccessTokenCacheItem cachedAccessTokenItem)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.<ExecuteAsync>d__2.MoveNext()
**Which version of MSAL.NET are you using?**
MSAL.NET 4.32.1
```

After digging into our internal logs, we found that the issue had to with the auto region discovery and how MSAL.NET handles the client assertion when creating a token request. More details below

**Tl;Dr:**
Audience claim contains the auto discovered region on oauth request as follows (but it should not) `https://**usgovvirginia**.login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token`

**Platform**
.netcore3.1, but framework does not matter here. The same issue repros on .net46, .net47 and .netcore2.1 

**What authentication flow has the issue?**
Confidential Client auth flow with ESTS-R enabled via `WithAzureRegion()` construct on ConfidentialClientApplicationBuilder.

**Is this a new or existing app?**
The app is in production, and I have upgraded to a new version of MSAL.

**Repro**

```csharp
        [TestMethod]
        public async Task MSAL_AutoRegionDiscoveryProblem()
        {
            X509Certificate2 cert;
            using (var store = new X509Store(StoreName.My, StoreLocation.LocalMachine))
            {
                store.Open(OpenFlags.ReadOnly);
                cert = store.Certificates.Find(X509FindType.FindBySubjectDistinguishedName, "<your subject distinguished name", validOnly: false)
                    .OfType<X509Certificate2>().FirstOrDefault() ?? throw new InvalidOperationException("No cert");
            }

            var cca = ConfidentialClientApplicationBuilder
                .Create("<your client id>")
                .WithAuthority("https://login.microsoftonline.com", Guid.Parse("72f988bf-86f1-41af-91ab-2d7cd011db47"), validateAuthority: true)
                .WithCertificate(cert)
                .WithHttpClientFactory(new SillyMsalClientFactory()) // this is where the repro magic happens
                .WithAzureRegion("TryAutoDetect") // problem
                .Build();

            await cca.AcquireTokenForClient(new[] { "https://graph.windows.net/.default" })
                .WithSendX5C(true)
                .ExecuteAsync(CancellationToken.None);
        }

        private class SillyMsalClientFactory : IMsalHttpClientFactory
        {
            private readonly HttpClient httpClient;

            public SillyMsalClientFactory()
            {
                this.httpClient = new HttpClient(new SillyMessageHandler());
            }

            public HttpClient GetHttpClient()
            {
                return this.httpClient;
            }
        }

        private class SillyMessageHandler : HttpClientHandler
        {
            private readonly string ImdsDiscoverRegionUrl = "http://169.254.169.254/metadata/instance/compute/location";

            public SillyMessageHandler()
            {
                UseCookies = false; // not important in this scenario
            }

            protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            {
                Console.WriteLine($"[{nameof(SillyMessageHandler)}] begin request to [{request.RequestUri}]");

                HttpResponseMessage resp;
                if (string.Equals(request.RequestUri.GetLeftPart(UriPartial.Path), ImdsDiscoverRegionUrl, StringComparison.Ordinal))
                {
                    // mocking region response from IMDS
                    resp = new HttpResponseMessage
                    {
                        StatusCode = System.Net.HttpStatusCode.OK,
                        RequestMessage = request,
                        Content = new ByteArrayContent(Encoding.UTF8.GetBytes("usgovvirginia")) // valid region in Azure Gov but not in Public
                    };
                    resp.Content.Headers.Add("content-type", "text/plain");
                }
                else if (string.Equals(request.RequestUri.AbsolutePath, "/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token", StringComparison.Ordinal))
                {
                    // Extract client assertion from oauth request
                    var handler = new JwtSecurityTokenHandler();

                    var reqBody = await request.Content.ReadAsStringAsync();
                    var reqBodyParsed = System.Web.HttpUtility.ParseQueryString(reqBody);

                    var clientAssertionJwt = reqBodyParsed.Get("client_assertion");
                    var clientAssertionJwtDecoded = handler.ReadJwtToken(clientAssertionJwt);

                    Console.WriteLine($"[{nameof(SillyMessageHandler)}] aud claim of the client assertion payload -- {clientAssertionJwtDecoded.Claims.Single(c => c.Type == "aud")}");

                    // Send request
                    resp = await base.SendAsync(request, cancellationToken);

                    // Log the full response from ESTS on failure
                    if (resp.StatusCode != System.Net.HttpStatusCode.OK)
                    {
                        var respBody = await resp.Content.ReadAsStringAsync();
                        Console.WriteLine($"[{nameof(SillyMessageHandler)}] response body: {respBody}");
                    }
                }
                else
                {
                    resp = await base.SendAsync(request, cancellationToken);
                }

                Console.WriteLine($"[{nameof(SillyMessageHandler)}] end request to [{request.RequestUri}] -- status code {resp.StatusCode}");
                return resp;
            }
        }
```

IF the `WithAzureRegion()` is present, then logs are as follows
```
    [SillyMessageHandler] begin request to [http://169.254.169.254/metadata/instance/compute/location?api-version=2020-06-01&format=text]
    [SillyMessageHandler] end request to [http://169.254.169.254/metadata/instance/compute/location?api-version=2020-06-01&format=text] -- status code OK
    [SillyMessageHandler] begin request to [https://usgovvirginia.login.microsoft.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token]
 ***PROBLEM***   [SillyMessageHandler] aud claim of the client assertion payload -- aud: https://usgovvirginia.login.microsoft.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token
    [SillyMessageHandler] response body: {"error":"invalid_client","error_description":"AADSTS700023: Client assertion audience claim does not match Realm issuer. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .\r\nTrace ID: 3ae3ffa3-1652-42f7-87a3-02f8d838b602\r\nCorrelation ID: 139d9ce4-da12-4555-b0ea-a974718ba101\r\nTimestamp: 2021-07-06 19:39:33Z","error_codes":[700023],"timestamp":"2021-07-06 19:39:33Z","trace_id":"3ae3ffa3-1652-42f7-87a3-02f8d838b602","correlation_id":"139d9ce4-da12-4555-b0ea-a974718ba101"}
    [SillyMessageHandler] end request to [https://usgovvirginia.login.microsoft.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token] -- status code Unauthorized
```

WITHOUT `WithAzureRegion()` the logs are as follows
```
    [SillyMessageHandler] begin request to [https://login.microsoftonline.com/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2F72f988bf-86f1-41af-91ab-2d7cd011db47%2Foauth2%2Fv2.0%2Fauthorize]
    [SillyMessageHandler] end request to [https://login.microsoftonline.com/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2F72f988bf-86f1-41af-91ab-2d7cd011db47%2Foauth2%2Fv2.0%2Fauthorize] -- status code OK
    [SillyMessageHandler] begin request to [https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token]
    [SillyMessageHandler] aud claim of the client assertion payload -- aud: https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token
    [SillyMessageHandler] end request to [https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/oauth2/v2.0/token] -- status code OK

```

**Expected behavior**
MSAL.NET should not append the discovered region to the aud claim on the client assertion payload

**Actual behavior**
MSAL.NET appends the discovered region to aud claim to client assertion payload

**Possible solution**
I may be wrong, but I believe the issue is coming from this line of code 

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/3f8c2c96b28eec5576933dc0f1ab4c7338475d79/src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialHelper.cs#L75

Where `endpoints.SelfSignedJwtAudience` is set to `tokenEndpoint` https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/0d53f3e80e29f9c6d172e069c10e02d71e24f65e/src/client/Microsoft.Identity.Client/Instance/AuthorityEndpoints.cs#L12

which in turn is constructed here https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/0d53f3e80e29f9c6d172e069c10e02d71e24f65e/src/client/Microsoft.Identity.Client/Instance/Discovery/RegionDiscoveryProvider.cs#L37 when regional discovery is enabled.

The fix here would be to have the *SelfSignedJwtAudience* match the non-regional authority endpoint.

**Comment by jmprieur** [comment]
@agolotin : do you think it's a regression?
@neha-bhargava can you please have a look?
cc: @henrik-me

**Comment by agolotin** [comment]
I am unsure whether this is a regression since 4.29.0 when IMDS based region discovery was introduced. The issue could have existed previously, I have not dug into the code before 4.29.

**Comment by bgavrilMS** [comment]
Regional is not supported on these clouds yet. Please contact Randeep for details.
